### PR TITLE
Improve performance by pooling dictionaries, sets, analysis data inst…

### PR DESCRIPTION
**Pull request by mavasani**
…ances, etc., which were showing up in performance traces.

Also made all the analysis data tracked by the analyses to be disposable, and the core dictionary based data are now backed by a pooled dictionary, which is released back to the pool on dispose.

### CODE REVIEW

File: src/Utilities/FlowAnalysis/Framework/DataFlow/AbstractAnalysisData.cs [hunk 1]
```diff
@@ -0,0 +1,26 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+
+namespace Microsoft.CodeAnalysis.FlowAnalysis.DataFlow
+{
+    internal abstract class AbstractAnalysisData : IDisposable
+    {
+        public bool Disposed { get; private set; }
```
Comment (dotpaul) [inline src/Utilities/FlowAnalysis/Framework/DataFlow/AbstractAnalysisData.cs]: nit: IsDisposed?  Or is Disposed more conventional?

**Comment by mavasani** [comment]
retest prtest/Windows/NT/Debug please

File: src/Utilities/FlowAnalysis/Framework/DataFlow/AbstractAnalysisData.cs (hunk 1 above)
Comment (mavasani) [reply src/Utilities/FlowAnalysis/Framework/DataFlow/AbstractAnalysisData.cs]: Sure, can change.

File: src/Utilities/FlowAnalysis/Framework/DataFlow/SetAbstractDomain.cs [hunk 2]
```diff
@@ -23,7 +23,9 @@ public override int Compare(ImmutableHashSet<T> oldValue, ImmutableHashSet<T> ne
             }
 
             int result;
-            var isSubset = oldValue.IsSubsetOf(newValue);
+            // PERF: Avoid additional hash set allocation by using overload which takes
```
Comment (dotpaul) [inline src/Utilities/FlowAnalysis/Framework/DataFlow/SetAbstractDomain.cs]: nit: comment doesn't match code

Comment (mavasani) [reply src/Utilities/FlowAnalysis/Framework/DataFlow/SetAbstractDomain.cs]: The new overload (extension method) used does take immutable hash set as the argument. Am I missing something?

File: src/Utilities/FlowAnalysis/Framework/DataFlow/AnalysisEntityDataFlowOperationVisitor.cs [hunk 3]
```diff
@@ -270,45 +272,41 @@ private void TransferValueTypeInstanceAnalysisDataForAssignment(AnalysisEntity t
...
-            var hasValueCopySemantics = analysisEntity.Type.HasValueCopySemantics();
-            foreach (var entity in GetChildAnalysisEntities(analysisEntity.InstanceLocation))
+            if (analysisEntity.InstanceLocation == null)
             {
-                if (!hasValueCopySemantics || entity.HasAncestor(analysisEntity))
-                {
-                    yield return entity;
-                }
+                return ImmutableHashSet<AnalysisEntity>.Empty;
+
```
[the remark this answers is not in the document]
Comment (mavasani) [reply src/Utilities/FlowAnalysis/Framework/DataFlow/AnalysisEntityDataFlowOperationVisitor.cs]: Thanks, will fix in next change.

File: src/Utilities/FlowAnalysis/Framework/DataFlow/SetAbstractDomain.cs (hunk 2 above)
Comment (dotpaul) [reply src/Utilities/FlowAnalysis/Framework/DataFlow/SetAbstractDomain.cs]: Oops, sorry, nevermind, didn't realize the existing `IsSubsetOf()` did instantiate a new HashSet.

**Pull request merged**